content/courses/advanced-ada/parts/resource_management/anonymous_access_types.rst

```diff
@@ -1958,12 +1960,559 @@ access-to-subprogram types can be solved by declaring them as anonymous types.
+            := F;
+       F2 : access function (B : Float)
+                             return Integer
+            := F1 (2);
+       I  : Integer := F2 (0.1);
+    begin
+       I := F1 (2) (0.1);
+    end Show_Readability_Issue;
+
+Therefore, our recommendation is to avoid this kind of *access cascading* by
+carefully designing your application. In general, you won't need that.
+
+
 .. _Adv_Ada_Accessibility_Rules_Anonymous_Access_To_Subprograms:
 
 Accessibility Rules and Anonymous Access-To-Subprograms
 -------------------------------------------------------
 
-.. todo::
+In principle, the
+:ref:`accessibility rules for anonymous access types <Adv_Ada_Anonymous_Access_Types_Accessibility_Rules>`
+that we've seen before apply to anonymous access-to-subprograms as well. Also,
+we had a discussion about
+:ref:`accessibility rules and access-to-subprograms <Adv_Ada_Accessibility_Rules_Access_To_Subprograms>`
+in the previous chapter. In this section, we review some of the rules that we
+already know and discuss how they relate to anonymous access-to-subprograms.
 
-    Complete section!
+.. admonition:: In the Ada Reference Manual
+
+    - :arm22:`3.10 Access Types <3-10>`
+
+
+Named vs. anonymous access-to-subprograms
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's see an example of a named access-to-subprogram type:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Anonymous_Access_Types.Accessibility_Rules_Anonymous_Access_To_Subprograms.Simple_Example_Named
+    :class: ada-expect-compile-error
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Access_To_Subprogram_Error is
+
+       type PI is access
+         procedure (I : in out Integer);
+
+       P : PI;
+
+       I : Integer := 0;
+    begin
+       declare
+          procedure Add_One (I : in out Integer) is
+          begin
+             I := I + 1;
+          end Add_One;
+       begin
+          P := Add_One'Access;
+       end;
+    end Show_Access_To_Subprogram_Error;
+
+In this example, we get a compilation error because the lifetime of the
+:ada:`Add_One` procedure is shorter than the access type :ada:`PI`.
+
+In contrast, using an anonymous access-to-subprogram type eliminates the
+compilation error, i.e. the assignment :ada:`P := Add_One'Access` becomes
+legal:
+
+.. code:: ada run_button project=Courses.Advanced_Ada.Resource_Management.Anonymous_Access_Types.Accessibility_Rules_Anonymous_Access_To_Subprograms.Simple_Example_Anonymous
+    :class: ada-run-expect-failure
+
+    with Ada.Text_IO; use Ada.Text_IO;
+
+    procedure Show_Access_To_Subprogram_Error is
+       P : access procedure (I : in out Integer);
+
+       I : Integer := 0;
+    begin
+       declare
+          procedure Add_One (I : in out Integer) is
+          begin
+             I := I + 1;
+          end Add_One;
+       begin
+          P := Add_One'Access;
+          --  ERROR: Add_One is out-of-scope after
```

Comment:
The comment here, indicating an error, suggests that the assignment is illegal, at compile-time. (At least to me.) Maybe change "ERROR" to "RUNTIME ERROR" ?  I see that the paragraph that follows says that it is an error at runtime, so maybe it doesn't need the change.

Reply:
Yes, I agree, it might be confusing. I'll add the word "RUNTIME" to avoid confusion.